### ISSUE TITLE
When Steam starts the browser to open a link from a message, it blocks.

### ISSUE THREAD

**Issue by xorgy** [ValveSoftware/steam-for-linux#5090]
#### Your system information

* Steam client version (build number or date): 2017-07-17T23:10:56Z (Steam package versions: 1500335472)
* Distribution (e.g. Ubuntu): Arch Linux
* Opted into Steam client beta?: No
* Have you checked for system updates?: Yes

#### Please describe your issue in as much detail as possible:
I clicked on a link in a message with no Chrome window open, and when the window opened, Steam froze (blocked) entirely. When I killed all the Chrome processes, Steam resumed.

#### Steps for reproducing this issue:

1. Click a link in a Steam message with no Chrome window already open.

**Comment by Spanguole** [comment]
I have the same issue with firefox.  
Also happens if steam opens [whatever the default file manager is], e.g. when clicking "show on disk" in a screenshot library.

**Comment by FreeSlave** [comment]
Probably Steam should use double-fork technique here

**Comment by xorgy** [comment]
@FreeSlave Yeah, in my case I'm not using a desktop environment. I have a thin window manager running on xinit. It should definitely double fork, and I think Steam does this correctly elsewhere, but perhaps just not in chat.

**Comment by xorgy** [comment]
Oh.. nevermind, it has the same problem when opening the link in the Help > About window.

**Comment by af7567** [comment]
This is the same as my issue #4727
I have just had this issue occur when opening a link for the first time, previously it only happened when steam opened the file manager.
I am using Steam beta from Dec 14 2018, firefox as the browser, thunar as the file manager.
If firefox is already running then when steam executes xdg-run it will tell the existing firefox process to open the new tab and everything is fine.
If firefox is not already running then steam will start it via xdg-run but then get stuck until I exit firefox.

**Comment by raingloom** [comment]
Having the same issue for quite some time now.

**Comment by RunningDroid** [comment]
I've been experiencing this bug on Void so made a workaround: https://github.com/RunningDroid/steam-xdg-open

I use it by launching a script that changes my $PATH before exec'ing /usr/bin/steam but it's probably safe to just add it to the $PATH for your user.

**Comment by af7567** [comment]
I forgot about this bug. just had it happen again when I clicked a link on a users profile which opened firefox via xdg-open. I saw @RunningDroid fix for it by changing xdg-open, but is this an xdg-open bug or a steam client one?

**Comment by RunningDroid** [comment]
> I forgot about this bug. just had it happen again when I clicked a link on a users profile which opened firefox via xdg-open. I saw @RunningDroid fix for it by changing xdg-open, but is this an xdg-open bug or a steam client one?

That's a matter of opinion; it could be argued that it's a bug in xdg-open because it could double-fork (if it weren't written in shell) and avoid blocking the parent process, but it could also be argued that it's a Steam client bug because it doesn't work around xdg-open not double-forking (and breaks the user's expectations by freezing when they click on a link.)

**Comment by xorgy** [comment]
@RunningDroid Given that essentially zero other applications have this issue, xdg-open not double-forking doesn't seem like the bug.

**Comment by sparr** [comment]
Per #7419, I'll mention this is also a problem with `Browse Local Files` for an installed game if xdg-open blocks when launching the file manager.

**Comment by sparr** [comment]
Discussion of this behavior and why it exists here: https://unix.stackexchange.com/questions/74605/use-xdg-open-to-open-a-url-with-a-new-process

**Comment by bladefox1337** [comment]
```
xdg-dbus-proxy "$DBUS_SESSION_BUS_ADDRESS" "$XDG_RUNTIME_DIR/steam-bus-proxy" --filter --call="org.freedesktop.DBus.*=*" --call="org.freedesktop.IBus.*=*" --call="org.freedesktop.portal.*=*" --call="com.feralinteractive.GameMode.*=*" --call="com.canonical.*=*" --broadcast="*=*" --call="org.kde.*=*" --call="org.gtk.*=*" --call="org.Nemo.*=*" --call="org.mozilla.firefoxdeveloperedition.*=*"&

export DBUS_SESSION_BUS_ADDRESS="unix:path=$XDG_RUNTIME_DIR/steam-bus-proxy"
```

thanks https://github.com/ValveSoftware/steam-for-linux/issues/5607#issuecomment-735312261, after adding ```--call="org.mozilla.firefoxdeveloperedition.*=*"```, i don't have any problems, without that i get message
```Firefox is already running, but is not responding. To use Firefox, you must first close the existing Firefox process, restart your device, or use a different profile.```, and steam get freezed, for unfreeze need close firefox.

**Comment by af7567** [comment]
It looks like some work has been done to fix this in beta, but it's still not right. Steam no longer "freezes" for me when clicking a link which spawns the browser, or when clicking "show on disk" for the screenshots.

But, clicking a link or button a second time won't work until quitting the browser which spawned the first time. I found this out after clicking a link multiple times and wondering why it wasn't opening - after closing firefox it reopened with the next link and then continued to reopen firefox for every time I had clicked it.

So it looks like the link handling has been moved to another thread which works through a queue so the UI doesn't get stuck, but this thread is getting stuck. It also seems to be impossible to quit steam while anything is in the queue.
